File: src/core/operations/pulls.ts
/**
 * Pull Request 相关的 Gitea API 操作。
 */
import type { GiteaClient } from '../giteaClient';
import type {
  GiteaChangedFile,
  GiteaIssueState,
  GiteaListResult,
  GiteaPullRequest,
  GiteaPullReview,
} from '../types';

/** PR 列表查询参数。 */
export interface ListPullsOptions {
  owner: string;
  repo: string;
  state?: GiteaIssueState;
  /** 排序：`oldest` / `recentupdate` / `leastupdate` / `mostcomment` / `leastcomment` / `priority`。 */
  sort?: string;
  /** 源分支过滤。 */
  head?: string;
  /** 目标分支过滤。 */
  base?: string;
  labels?: string;
  milestone?: number;
  page?: number;
  limit?: number;
}

/** 创建 PR 参数。 */
export interface CreatePullOptions {
  title: string;
  /** 源分支，可为 `owner:branch` 形式。 */
  head: string;
  /** 目标分支。 */
  base: string;
  body?: string;
  assignees?: string[];
  reviewers?: string[];
  teamReviewers?: string[];
  labels?: number[];
  milestone?: number;
  allowMaintainerEdit?: boolean;
}

/** 合并 PR 参数。 */
export interface MergePullOptions {
  /** 合并方式，默认 `merge`。 */
  strategy?: 'merge' | 'rebase' | 'rebase-merge' | 'squash' | 'fast-forward-only' | 'manually-merged';
  /** 合并后删除源分支。 */
  deleteBranchAfterMerge?: boolean;
  /** 强制合并（忽略分支保护）。 */
  forceMerge?: boolean;
  /** 自定义合并标题。 */
  mergeTitle?: string;
  /** 自定义合并提交信息。 */
  mergeMessage?: string;
  /** 校验通过后自动合并。 */
  mergeWhenChecksSucceed?: boolean;
  /** 目标提交 SHA，用于防止并发变更。 */
  headCommitId?: string;
}

/** 评审 PR 参数。 */
export interface ReviewPullOptions {
  /** 评审动作。 */
  event: 'APPROVED' | 'COMMENT' | 'REQUEST_CHANGES';
  /** 评审正文。 */
  body?: string;
  /** 关联的提交 SHA，缺省由服务端取最新。 */
  commitId?: string;
}

/** Pull Request 操作集合。 */
export class PullOperations {
  constructor(private readonly client: GiteaClient) {}

  /**
   * 列出 PR。
   * @param options 查询参数
   * @returns PR 列表
   */
  public async list(options: ListPullsOptions): Promise<GiteaListResult<GiteaPullRequest>> {
    const result = await this.client.requestWithMeta<GiteaPullRequest[]>(
      'GET',
      `/repos/${enc(options.owner)}/${enc(options.repo)}/pulls`,
      {
        query: {
          state: options.state ?? 'open',
          sort: options.sort,
          head: options.head,
          base: options.base,
          labels: options.labels,
          milestone: options.milestone,
          page: options.page ?? 1,
          limit: options.limit ?? 50,
        },
      },
    );
    return { items: result.data ?? [], pageInfo: result.pageInfo };
  }

  /**
   * 获取单个 PR。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index PR 序号
   * @returns PR 详情
   */
  public get(owner: string, repo: string, index: number): Promise<GiteaPullRequest> {
    return this.client.request<GiteaPullRequest>('GET', `/repos/${enc(owner)}/${enc(repo)}/pulls/${index}`);
  }

  /**
   * 创建 PR。
   * @param owner 所属者
   * @param repo 仓库名
   * @param options 创建参数
   * @returns 新建的 PR
   */
  public create(owner: string, repo: string, options: CreatePullOptions): Promise<GiteaPullRequest> {
    const body: Record<string, unknown> = {
      title: options.title,
      head: options.head,
      base: options.base,
    };
    if (options.body !== undefined) body.body = options.body;
    if (options.assignees?.length) body.assignees = options.assignees;
    if (options.reviewers?.length) body.reviewers = options.reviewers;
    if (options.teamReviewers?.length) body.team_reviewers = options.teamReviewers;
    if (options.labels?.length) body.labels = options.labels;
    if (options.milestone !== undefined) body.milestone = options.milestone;
    if (options.allowMaintainerEdit !== undefined) body.allow_maintainer_edit = options.allowMaintainerEdit;
    return this.client.request<GiteaPullRequest>('POST', `/repos/${enc(owner)}/${enc(repo)}/pulls`, { body });
  }

  /**
   * 获取 PR 的文本差异（unified diff）。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index PR 序号
   * @param format `diff` 或 `patch`
   * @returns diff 文本
   */
  public getDiff(
    owner: string,
    repo: string,
    index: number,
    format: 'diff' | 'patch' = 'diff',
  ): Promise<string> {
    return this.client.request<string>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/pulls/${index}.${format}`,
      { responseType: 'text' },
    );
  }

  /**
   * 获取 PR 变更文件清单。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index PR 序号
   * @returns 变更文件列表
   */
  public async listFiles(owner: string, repo: string, index: number): Promise<GiteaChangedFile[]> {
    const result = await this.client.requestWithMeta<GiteaChangedFile[]>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/pulls/${index}/files`,
      { query: { limit: 300 } },
    );
    return result.data ?? [];
  }

  /**
   * 合并 PR。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index PR 序号
   * @param options 合并参数
   * @returns 合并结果（服务端可能返回空响应体）
   */
  public async merge(
    owner: string,
    repo: string,
    index: number,
    options: MergePullOptions = {},
  ): Promise<{ merged: boolean; message: string }> {
    const body: Record<string, unknown> = { do: options.strategy ?? 'merge' };
    if (options.deleteBranchAfterMerge !== undefined) {
      body.delete_branch_after_merge = options.deleteBranchAfterMerge;
    }
    if (options.forceMerge !== undefined) body.force_merge = options.forceMerge;
    if (options.mergeTitle) body.merge_title_field = options.mergeTitle;
    if (options.mergeMessage) body.merge_message_field = options.mergeMessage;
    if (options.mergeWhenChecksSucceed !== undefined) {
      body.merge_when_checks_succeed = options.mergeWhenChecksSucceed;
    }
    if (options.headCommitId) body.head_commit_id = options.headCommitId;

    await this.client.request('POST', `/repos/${enc(owner)}/${enc(repo)}/pulls/${index}/merge`, { body });
    return { merged: true, message: `PR #${index} 已按 ${String(body.do)} 方式合并` };
  }

  /**
   * 更新 PR（标题 / 正文 / 状态 / 指派 / 目标分支）。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index PR 序号
   * @param options 更新参数
   * @returns 更新后的 PR
   */
  public update(
    owner: string,
    repo: string,
    index: number,
    options: {
      title?: string;
      body?: string;
      state?: 'open' | 'closed';
      assignees?: string[];
      base?: string;
    },
  ): Promise<GiteaPullRequest> {
    const body: Record<string, unknown> = {};
    if (options.title !== undefined) body.title = options.title;
    if (options.body !== undefined) body.body = options.body;
    if (options.state !== undefined) body.state = options.state;
    if (options.assignees !== undefined) body.assignees = options.assignees;
    if (options.base !== undefined) body.base = options.base;
    return this.client.request<GiteaPullRequest>(
      'PATCH',
      `/repos/${enc(owner)}/${enc(repo)}/pulls/${index}`,
      { body },
    );
  }

  /**
   * 列出 PR 的评审记录。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index PR 序号
   * @returns 评审列表
   */
  public async listReviews(owner: string, repo: string, index: number): Promise<GiteaPullReview[]> {
    const result = await this.client.requestWithMeta<GiteaPullReview[]>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/pulls/${index}/reviews`,
      { query: { limit: 100 } },
    );
    return result.data ?? [];
  }

  /**
   * 提交 PR 评审（批准 / 请求修改 / 评论）。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index PR 序号
   * @param options 评审参数
   * @returns 评审结果
   */
  public createReview(
    owner: string,
    repo: string,
    index: number,
    options: ReviewPullOptions,
  ): Promise<GiteaPullReview> {
    const body: Record<string, unknown> = { event: options.event };
    if (options.body !== undefined) body.body = options.body;
    if (options.commitId) body.commit_id = options.commitId;
    return this.client.request<GiteaPullReview>(
      'POST',
      `/repos/${enc(owner)}/${enc(repo)}/pulls/${index}/reviews`,
      { body },
    );
  }

}

/** URL 路径片段编码。 */
function enc(value: string): string {
  return encodeURIComponent(value);
}

File: python/src/gitea_toolkit_mcp/operations/pulls.py
"""Pull Request 域操作（移植自 TS 版 `src/core/operations/pulls.ts`）。"""

from __future__ import annotations

from typing import Any, Literal
from urllib.parse import quote

from ..client import GiteaClient, ListResult

#: 排序方式；取值与 Gitea 的 ``sort`` 查询参数一致。
PullSort = Literal[
    "oldest",
    "recentupdate",
    "leastupdate",
    "mostcomment",
    "leastcomment",
    "priority",
]

#: 合并方式；与 Gitea 的 ``do`` 字段一致。
MergeStrategy = Literal[
    "merge",
    "rebase",
    "rebase-merge",
    "squash",
    "fast-forward-only",
    "manually-merged",
]

#: 评审动作。
ReviewEvent = Literal["APPROVED", "COMMENT", "REQUEST_CHANGES"]


def _enc(value: str) -> str:
    """URL 路径片段编码。"""
    return quote(str(value), safe="")


class PullOperations:
    """PR 的查询、创建、合并与评审。"""

    def __init__(self, client: GiteaClient) -> None:
        self._client = client

    async def list(
        self,
        owner: str,
        repo: str,
        *,
        state: str = "open",
        sort: PullSort | None = None,
        head: str | None = None,
        base: str | None = None,
        labels: str | None = None,
        milestone: int | None = None,
        page: int = 1,
        limit: int = 50,
    ) -> ListResult:
        """列出 PR。

        走 request_paged：服务端单页上限为 max_response_items（默认 50），
        想要更多必须真的翻页，只调大 limit 会被静默截断。
        """
        return await self._client.request_paged(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/pulls",
            limit=limit,
            start_page=page,
            query={
                "state": state,
                "sort": sort,
                "head": head,
                "base": base,
                "labels": labels,
                "milestone": milestone,
            },
            extract=lambda data: data if isinstance(data, list) else [],
        )

    async def get(self, owner: str, repo: str, index: int) -> dict[str, Any]:
        """取单个 PR 详情。"""
        return await self._client.request(
            "GET", f"/repos/{_enc(owner)}/{_enc(repo)}/pulls/{index}"
        )

    async def create(
        self,
        owner: str,
        repo: str,
        *,
        title: str,
        head: str,
        base: str,
        body: str | None = None,
        assignees: list[str] | None = None,
        reviewers: list[str] | None = None,
        team_reviewers: list[str] | None = None,
        labels: list[int] | None = None,
        milestone: int | None = None,
        allow_maintainer_edit: bool | None = None,
    ) -> dict[str, Any]:
        """创建 PR。

        ``head`` 允许写成 ``owner:branch``，用于从 fork 发起跨仓库 PR。
        """
        payload: dict[str, Any] = {"title": title, "head": head, "base": base}
        if body is not None:
            payload["body"] = body
        if assignees:
            payload["assignees"] = assignees
        if reviewers:
            payload["reviewers"] = reviewers
        if team_reviewers:
            payload["team_reviewers"] = team_reviewers
        if labels:
            payload["labels"] = labels
        if milestone is not None:
            payload["milestone"] = milestone
        if allow_maintainer_edit is not None:
            payload["allow_maintainer_edit"] = allow_maintainer_edit
        return await self._client.request(
            "POST", f"/repos/{_enc(owner)}/{_enc(repo)}/pulls", body=payload
        )

    async def get_diff(
        self,
        owner: str,
        repo: str,
        index: int,
        *,
        format: Literal["diff", "patch"] = "diff",  # noqa: A002 - 与工具层参数名保持一致
    ) -> str:
        """取 PR 的文本差异。

        ``.diff`` / ``.patch`` 返回的是**纯文本**，必须走 ``response_type="text"``
        ——响应不是 JSON，按 JSON 解析会直接失败。
        """
        raw = await self._client.request(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/pulls/{index}.{format}",
            response_type="text",
        )
        return raw if isinstance(raw, str) else ""

    async def list_files(self, owner: str, repo: str, index: int) -> list[Any]:
        """列出 PR 变更文件。

        用 ``request_paged``（上限放到 300）：单次请求在改动文件多的 PR 上
        会被服务端截断到 50 个且不报错，评审时会漏掉后面的文件。
        """
        result = await self._client.request_paged(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/pulls/{index}/files",
            limit=300,
            extract=lambda data: data if isinstance(data, list) else [],
        )
        return list(result.items)

    async def merge(
        self,
        owner: str,
        repo: str,
        index: int,
        *,
        strategy: MergeStrategy | None = None,
        delete_branch_after_merge: bool | None = None,
        force_merge: bool | None = None,
        merge_title: str | None = None,
        merge_message: str | None = None,
        merge_when_checks_succeed: bool | None = None,
        head_commit_id: str | None = None,
    ) -> dict[str, Any]:
        """合并 PR。

        服务端成功时可能返回**空响应体**，所以成功与否按「有没有抛异常」判定，
        返回文案由本地拼装（与 TS 版一致）。

        字段名按实例 swagger 的 ``MergePullRequestOption`` 核对：
        **全是小写蛇形**（``do`` / ``merge_title_field`` / ``merge_message_field``），
        写成 Go 结构体的 ``Do`` / ``MergeTitleField`` 会被服务端判为缺字段而报错。
        """
        strategy_value = strategy or "merge"
        payload: dict[str, Any] = {"do": strategy_value}
        if delete_branch_after_merge is not None:
            payload["delete_branch_after_merge"] = delete_branch_after_merge
        if force_merge is not None:
            payload["force_merge"] = force_merge
        if merge_title:
            payload["merge_title_field"] = merge_title
        if merge_message:
            payload["merge_message_field"] = merge_message
        if merge_when_checks_succeed is not None:
            payload["merge_when_checks_succeed"] = merge_when_checks_succeed
        if head_commit_id:
            payload["head_commit_id"] = head_commit_id
        await self._client.request(
            "POST",
            f"/repos/{_enc(owner)}/{_enc(repo)}/pulls/{index}/merge",
            body=payload,
        )
        return {"merged": True, "message": f"PR #{index} 已按 {strategy_value} 方式合并"}

    async def create_review(
        self,
        owner: str,
        repo: str,
        index: int,
        *,
        event: ReviewEvent,
        body: str | None = None,
        commit_id: str | None = None,
    ) -> dict[str, Any]:
        """提交 PR 评审（批准 / 请求修改 / 仅评论）。"""
        payload: dict[str, Any] = {"event": event}
        if body is not None:
            payload["body"] = body
        if commit_id:
            payload["commit_id"] = commit_id
        return await self._client.request(
            "POST",
            f"/repos/{_enc(owner)}/{_enc(repo)}/pulls/{index}/reviews",
            body=payload,
        )


__all__ = ["MergeStrategy", "PullOperations", "PullSort", "ReviewEvent"]
